File: src/aexpy/models/description.py
from enum import IntEnum, IntFlag
from functools import cached_property
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

from ..utils import isPrivateName
from .typing import TypeType

TRANSFER_BEGIN = "AEXPY_TRANSFER_BEGIN"


class Location(BaseModel):
    file: str = ""
    line: int = -1
    module: str = ""

    def __str__(self, /):
        return f"{self.file}:{self.line}:{self.module}"


class ApiEntry(BaseModel):
    name: str = ""
    id: str = ""
    alias: list[str] = []
    docs: str = ""
    comments: str = ""
    src: str = ""
    location: Location | None = None
    private: bool = False
    deprecated: bool = False
    parent: str = ""
    data: dict[str, Any] = {}


class CollectionEntry(ApiEntry):
    members: dict[str, str] = {}
    slots: set[str] = set()
    annotations: dict[str, str] = {}

    @cached_property
    def aliasMembers(self, /):
        return {k: v for k, v in self.members.items() if v != f"{self.id}.{k}"}


class ItemScope(IntEnum):
    Static = 0
    Class = 1
    Instance = 2


class ItemEntry(ApiEntry):
    scope: ItemScope = ItemScope.Static
    type: Annotated[TypeType, Field(discriminator="form")] | None = None


class SpecialKind(IntEnum):
    Unknown = 0
    Empty = 1
    External = 2


class SpecialEntry(ApiEntry):
    form: Literal["special"] = "special"
    kind: SpecialKind = SpecialKind.Unknown


class ModuleEntry(CollectionEntry):
    form: Literal["module"] = "module"


class ClassFlag(IntFlag):
    Empty = 0
    Abstract = 1 << 0
    Final = 1 << 1
    Generic = 1 << 2
    Dataclass = 1 << 10


class ClassEntry(CollectionEntry):
    form: Literal["class"] = "class"

    bases: list[str] = []
    subclasses: list[str] = []
    abcs: list[str] = []
    mros: list[str] = []
    flags: ClassFlag = ClassFlag.Empty


class AttributeEntry(ItemEntry):
    form: Literal["attr"] = "attr"

    rawType: str = ""
    annotation: str = ""
    property: bool = False


class ParameterKind(IntEnum):
    Positional = 0
    PositionalOrKeyword = 1
    VarPositional = 2
    Keyword = 3
    VarKeyword = 4
    VarKeywordCandidate = 5


class Parameter(BaseModel):
    kind: ParameterKind = ParameterKind.PositionalOrKeyword
    name: str = ""
    annotation: str = ""
    default: str | None = None
    """Default value. None for variable default value."""
    optional: bool = False
    source: str = ""
    type: Annotated[TypeType, Field(discriminator="form")] | None = None

    @cached_property
    def isKeyword(self, /):
        return self.kind in {
            ParameterKind.Keyword,
            ParameterKind.PositionalOrKeyword,
            ParameterKind.VarKeywordCandidate,
        }

    @cached_property
    def isPositional(self, /):
        return self.kind in {
            ParameterKind.Positional,
            ParameterKind.PositionalOrKeyword,
        }

    @cached_property
    def isVar(self, /):
        return self.kind in {ParameterKind.VarKeyword, ParameterKind.VarPositional}


class FunctionFlag(IntFlag):
    Empty = 0
    Abstract = 1 << 0
    Final = 1 << 1
    Generic = 1 << 2
    Override = 1 << 10
    Async = 1 << 11
    TransmitKwargs = 1 << 20


class FunctionEntry(ItemEntry):
    form: Literal["func"] = "func"

    returnAnnotation: str = ""
    parameters: list[Parameter] = []
    annotations: dict[str, str] = {}
    returnType: Annotated[TypeType, Field(discriminator="form")] | None = None
    callers: list[str] = []
    callees: list[str] = []
    flags: FunctionFlag = FunctionFlag.Empty

    def getParameter(self, /, name: str):
        for p in self.parameters:
            if p.name == name:
                return p

    def position(self, /, parameter: Parameter):
        try:
            return self.positionals.index(parameter)
        except:
            return None

    @cached_property
    def positionalOnlys(self, /):
        return [x for x in self.parameters if x.kind == ParameterKind.Positional]

    @cached_property
    def positionals(self, /):
        return [x for x in self.parameters if x.isPositional]

    @cached_property
    def keywordOnlys(self, /):
        return [x for x in self.parameters if x.kind == ParameterKind.Keyword]

    @cached_property
    def keywords(self, /):
        return [x for x in self.parameters if x.isKeyword]

    @cached_property
    def candidates(self, /):
        return [
            x for x in self.parameters if x.kind == ParameterKind.VarKeywordCandidate
        ]

    @cached_property
    def varPositional(self, /):
        items = [x for x in self.parameters if x.kind == ParameterKind.VarPositional]
        if len(items) > 0:
            return items[0]
        return None

    @cached_property
    def varKeyword(self, /):
        items = [x for x in self.parameters if x.kind == ParameterKind.VarKeyword]
        if len(items) > 0:
            return items[0]
        return None


type ApiEntryType = ModuleEntry | ClassEntry | FunctionEntry | AttributeEntry | SpecialEntry


def isPrivate(entry: ApiEntry):
    names = [entry.id, *entry.alias]
    for alias in names:
        if not isPrivateName(alias):
            return False
    return True

File: src/web/src/models/description.ts
export class Location {
    file: string = "";
    line: number = -1;
    module: string = "";

    from(data: any) {
        this.file = data.file ?? "";
        this.line = data.line ?? -1;
        this.module = data.module ?? "";
        return this;
    }
}

export class ApiEntry {
    name: string = "";
    id: string = "";
    alias: string[] = [];
    docs: string = "";
    comments: string = "";
    src: string = "";
    location?: Location;
    private: boolean = false;
    deprecated: boolean = false;
    data: any = {};
    parent: string = "";

    from(data: any) {
        this.name = data.name ?? "";
        this.id = data.id ?? "";
        this.alias = data.alias ?? [];
        this.docs = data.docs ?? "";
        this.comments = data.comments ?? "";
        this.src = data.src ?? "";
        this.private = data.private ?? false;
        this.deprecated = data.deprecated ?? false;
        this.data = data.data ?? {};
        this.parent = data.parent ?? "";

        if (data.location) {
            this.location = new Location();
            this.location.from(data.location);
        }
        return this;
    }
}

export function getTypeColor(entry: ApiEntry | string) {
    if (entry instanceof ApiEntry) {
        entry = <string>Object.getPrototypeOf(entry).constructor.name.replace("Entry", "");
    }

    switch (entry) {
        case "Module":
            return '#2080f0';
        case "Class":
            return '#f0a020';
        case "Function":
            return '#18a058';
        case "Attribute":
            return '#d03050';
        default:
            return '#666666';
    }
}

export class CollectionEntry extends ApiEntry {
    members: { [key: string]: string } = {};
    annotations: { [key: string]: string } = {};
    slots: Set<string> = new Set<string>();

    from(data: any) {
        super.from(data);
        this.members = data.members ?? {};
        this.annotations = data.annotations ?? {};
        this.slots = new Set<string>(data.slots ?? []);
        return this;
    }
}

export enum ItemScope {
    Static = 0,
    Class = 1,
    Instance = 2,
}

export class ItemEntry extends ApiEntry {
    type?: any;
    scope: ItemScope = ItemScope.Static;


    from(data: any) {
        super.from(data);
        this.scope = data.scope ?? ItemScope.Static;
        this.type = data.type;
        return this;
    }
}

export enum SpecialKind {
    Unknown = 0,
    Empty = 1,
    External = 2,
}

export class SpecialEntry extends ApiEntry {
    kind: SpecialKind = SpecialKind.Unknown;
    data: string = "";

    from(data: any) {
        super.from(data);
        this.kind = data.kind ?? SpecialKind.Unknown;
        this.data = data.data ?? "";
        return this;
    }
}

export class ModuleEntry extends CollectionEntry {
}

export enum ClassFlag {
    Empty = 0,
    Abstract = 1 << 0,
    Final = 1 << 1,
    Generic = 1 << 2,
    Dataclass = 1 << 10,
}

export class ClassEntry extends CollectionEntry {
    bases: string[] = [];
    subclasses: string[] = [];
    abcs: string[] = [];
    mro: string[] = [];
    flags: ClassFlag = ClassFlag.Empty;

    from(data: any) {
        super.from(data);
        this.bases = data.bases ?? [];
        this.subclasses = data.subclasses ?? [];
        this.abcs = data.abcs ?? [];
        this.mro = data.mro ?? [];
        this.flags = data.flags ?? ClassFlag.Empty;
        return this;
    }
}

export class AttributeEntry extends ItemEntry {
    rawType: string = "";
    annotation: string = "";
    property: boolean = false;

    from(data: any) {
        super.from(data);
        this.rawType = data.rawType ?? "";
        this.annotation = data.annotation ?? "";
        this.property = data.property ?? false;
        return this;
    }
}

export enum ParameterKind {
    Positional = 0,
    PositionalOrKeyword = 1,
    VarPositional = 2,
    Keyword = 3,
    VarKeyword = 4,
    VarKeywordCandidate = 5,
}

export class Parameter {
    kind: ParameterKind = ParameterKind.PositionalOrKeyword;
    name: string = "";
    annotation: string = "";
    default?: string;
    optional: boolean = false;
    source: string = "";
    type?: any;

    from(data: any) {
        this.kind = data.kind ?? ParameterKind.PositionalOrKeyword;
        this.name = data.name ?? "";
        this.annotation = data.annotation ?? "";
        this.default = data.default;
        this.optional = data.optional ?? false;
        this.source = data.source ?? "";
        if (data.type != undefined) {
            this.type = data.type;
        }
        return this;
    }
}

export enum FunctionFlag {
    Empty = 0,
    Abstract = 1 << 0,
    Final = 1 << 1,
    Generic = 1 << 2,
    Override = 1 << 10,
    Async = 1 << 11,
    TransmitKwargs = 1 << 20,
}

export class FunctionEntry extends ItemEntry {
    returnAnnotation: string = "";
    parameters: Parameter[] = [];
    annotations: { [key: string]: string } = {};
    returnType?: any;
    callers: string[] = [];
    callees: string[] = [];
    flags: FunctionFlag = FunctionFlag.Empty;

    varPositional() {
        return this.parameters.find(p => p.kind == ParameterKind.VarPositional);
    }

    varKeyword() {
        return this.parameters.find(p => p.kind == ParameterKind.VarKeyword);
    }

    from(data: any) {
        super.from(data);
        this.returnAnnotation = data.returnAnnotation ?? "";
        (<any[]>data.parameters ?? []).forEach(element => {
            let para = new Parameter();
            para.from(element);
            this.parameters.push(para);
        });
        this.annotations = data.annotations ?? {};
        if (data.returnType != undefined) {
            this.returnType = data.returnType
        }
        this.callers = data.callers ?? [];
        this.callees = data.callees ?? [];
        this.flags = data.flags ?? FunctionFlag.Empty;
        return this;
    }
}

export function loadApiEntry(data: any): ApiEntry {
    let entry = undefined;
    switch (data.form) {
        case "attr":
            entry = new AttributeEntry();
            break;
        case "class":
            entry = new ClassEntry();
            break;
        case "func":
            entry = new FunctionEntry();
            break;
        case "module":
            entry = new ModuleEntry();
            break;
        case "special":
            entry = new SpecialEntry();
            break;
        default:
            throw new Error("Unknown schema: " + data.schema);
    }
    return entry.from(data);
}